### ISSUE TITLE
0.16.8 breaks aead::LessSafeKey

### ISSUE THREAD

**Issue by archseer** [briansmith/ring#895]
```
    let key = aead::UnboundKey::new(algorithm, key)?;
    let opening_key = aead::LessSafeKey::new(key);

```

results in

```
error[E0282]: type annotations needed for `ring::aead::LessSafeKey<N>`
   --> biscuit/src/jwa.rs:857:23
    |
857 |     let opening_key = aead::LessSafeKey::new(key);
    |         -----------   ^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for `N`
    |         |
    |         consider giving `opening_key` the explicit type `ring::aead::LessSafeKey<N>`, where the type parameter `N` is specified
```

Note that this is for 0.16.8 which was published yesterday, but no matching commits for it on github.

**Comment by jbg** [comment]
I’ve got projects suddenly breaking with this too. Pinning to `=0.16.7` (the previous release) temporarily solves the issue, but this breaking change in a minor release should probably be yanked?

**Comment by jbg** [comment]
In case anyone else’s spider sense also goes off when seeing software releases with no corresponding code in the repository; here’s the diff:

```
diff -ur ring/src/aead.rs ring-0.16.8/src/aead.rs
--- ring/src/aead.rs	2019-08-30 14:06:23.899988616 +0100
+++ ring-0.16.8/src/aead.rs	2019-08-28 04:26:13.000000000 +0100
@@ -22,7 +22,7 @@
 //! [`crypto.cipher.AEAD`]: https://golang.org/pkg/crypto/cipher/#AEAD
 
 use self::block::{Block, BLOCK_LEN};
-use crate::{constant_time, cpu, error, hkdf, polyfill};
+use crate::{constant_time, cpu, error, hkdf, polyfill, rand};
 use core::ops::RangeFrom;
 
 pub use self::{
@@ -445,18 +445,58 @@
     }
 }
 
+mod sealed {
+    pub trait NonceGeneration {
+        const VALUE: Self;
+    }
+}
+
+/// A nonce generation strategy for `LesssSafeKey`; either `ExplicitNonces` or
+/// `RandomNonces`.
+pub trait NonceGeneration: sealed::NonceGeneration {}
+impl<T> NonceGeneration for T where T: sealed::NonceGeneration {}
+
+/// A nonce generation strategy where random nonces are used.
+///
+/// These algorithms are *NOT* nonce-misuse-resistant. They use nonces small
+/// enough where birthday collisions need to be considered, so even ensuring
+/// perfectly-random generation of nonces isn't sufficient to prevent nonce
+/// reuse.
+pub struct RandomNonces(());
+impl sealed::NonceGeneration for RandomNonces {
+    const VALUE: Self = Self(());
+}
+
+/// A nonce generation strategy where explicitly-constructed nonces are used.
+pub struct ExplicitNonces(());
+impl sealed::NonceGeneration for ExplicitNonces {
+    const VALUE: Self = Self(());
+}
+
 /// Immutable keys for use in situations where `OpeningKey`/`SealingKey` and
 /// `NonceSequence` cannot reasonably be used.
 ///
-/// Prefer to use `OpeningKey`/`SealingKey` and `NonceSequence` when practical.
-pub struct LessSafeKey {
+/// These algorithms are *NOT* nonce-misuse-resistant. They use nonces small
+/// enough where the likelihood of collisions must be carefully considered.
+pub struct LessSafeKey<N = ExplicitNonces>
+where
+    N: NonceGeneration,
+{
     key: UnboundKey,
+    _nonce_generation: N,
 }
 
-impl LessSafeKey {
+impl<N> LessSafeKey<N>
+where
+    N: NonceGeneration,
+{
     /// Constructs a `LessSafeKey` from an `UnboundKey`.
+    #[inline]
     pub fn new(key: UnboundKey) -> Self {
-        Self { key }
+        Self {
+            key,
+            _nonce_generation: N::VALUE,
+        }
     }
 
     /// Like [`OpeningKey::open_in_place()`], except it accepts an arbitrary nonce.
@@ -492,6 +532,14 @@
         open_within_(&self.key, nonce, aad, in_out, ciphertext_and_tag)
     }
 
+    /// The key's AEAD algorithm.
+    #[inline]
+    pub fn algorithm(&self) -> &'static Algorithm {
+        &self.key.algorithm
+    }
+}
+
+impl LessSafeKey<ExplicitNonces> {
     /// Deprecated. Renamed to [`seal_in_place_append_tag()`].
     #[deprecated(note = "Renamed to `seal_in_place_append_tag`.")]
     #[inline]
@@ -523,6 +571,9 @@
         A: AsRef<[u8]>,
         InOut: AsMut<[u8]> + for<'in_out> Extend<&'in_out u8>,
     {
+        // Overwrite the plaintext with the ciphertext before extending `in_out`
+        // so that if the `extend()` causes a reallocation, the ciphertext (not
+        // the plaintext) will be in the old deallocated buffer.
         self.seal_in_place_separate_tag(nonce, aad, in_out.as_mut())
             .map(|tag| in_out.extend(tag.as_ref()))
     }
@@ -543,15 +594,63 @@
     {
         seal_in_place_separate_tag_(&self.key, nonce, Aad::from(aad.as_ref()), in_out)
     }
+}
 
-    /// The key's AEAD algorithm.
+impl LessSafeKey<RandomNonces> {
+    /// Like [`SealingKey::seal_in_place_append_tag()`], except the nonce is
+    /// randomly generated.
+    ///
+    /// The randomly-generated nonce is returned on success.
     #[inline]
-    pub fn algorithm(&self) -> &'static Algorithm {
-        &self.key.algorithm
+    pub fn seal_with_random_nonce_in_place_append_tag<A, InOut>(
+        &self,
+        aad: Aad<A>,
+        in_out: &mut InOut,
+        rng: &dyn rand::SecureRandom,
+    ) -> Result<Nonce, error::Unspecified>
+    where
+        A: AsRef<[u8]>,
+        InOut: AsMut<[u8]> + for<'in_out> Extend<&'in_out u8>,
+    {
+        // Overwrite the plaintext with the ciphertext before extending `in_out`
+        // so that if the `extend()` causes a reallocation, the ciphertext (not
+        // the plaintext) will be in the old deallocated buffer.
+        self.seal_with_random_nonce_in_place_separate_tag(aad, in_out.as_mut(), rng)
+            .map(|(nonce, tag)| {
+                in_out.extend(tag.as_ref());
+                nonce
+            })
+    }
+
+    /// Like [`SealingKey::seal_in_place_separate_tag()`], except the nonce is
+    /// randomly generated.
+    ///
+    /// The randomly-generated nonce and the tag are returned on success.
+    #[inline]
+    pub fn seal_with_random_nonce_in_place_separate_tag<A>(
+        &self,
+        aad: Aad<A>,
+        in_out: &mut [u8],
+        rng: &dyn rand::SecureRandom,
+    ) -> Result<(Nonce, Tag), error::Unspecified>
+    where
+        A: AsRef<[u8]>,
+    {
+        let nonce = Nonce::assume_unique_for_key(rand::generate(rng)?.expose());
+        seal_in_place_separate_tag_(
+            &self.key,
+            Nonce::assume_unique_for_key(*nonce.as_ref()),
+            Aad::from(aad.as_ref()),
+            in_out,
+        )
+        .map(|tag| (nonce, tag))
     }
 }
 
-impl core::fmt::Debug for LessSafeKey {
+impl<N> core::fmt::Debug for LessSafeKey<N>
+where
+    N: NonceGeneration,
+{
     fn fmt(&self, f: &mut core::fmt::Formatter) -> Result<(), core::fmt::Error> {
         f.debug_struct("LessSafeKey")
             .field("algorithm", self.algorithm())
diff -ur ring/src/hkdf.rs ring-0.16.8/src/hkdf.rs
--- ring/src/hkdf.rs	2019-08-30 14:06:23.906655463 +0100
+++ ring-0.16.8/src/hkdf.rs	2019-08-25 03:56:49.000000000 +0100
@@ -94,9 +94,9 @@
     }
 }
 
-/// The length of the OKM (Output Keying Material) for a `Prk::expand()` call.
+/// The length of the OKM (Output Keying Material) for a `Prf::expand()` call.
 pub trait KeyType {
-    /// The length that `Prk::expand()` should expand its input to.
+    /// The length that `Prf::expand()` should expand its input to.
     fn len(&self) -> usize;
 }
 
@@ -162,7 +162,7 @@
 }
 
 impl<L: KeyType> Okm<'_, L> {
-    /// The `OkmLength` given to `Prk::expand()`.
+    /// The `OkmLength` given to `Prf::expand()`.
     #[inline]
     pub fn len(&self) -> &L {
         &self.len
diff -ur ring/src/rand.rs ring-0.16.8/src/rand.rs
--- ring/src/rand.rs	2019-08-30 14:06:23.909988888 +0100
+++ ring-0.16.8/src/rand.rs	2019-08-28 04:17:29.000000000 +0100
@@ -97,7 +97,7 @@
         }
     }
 
-    impl_random_arrays![4 8 16 32 48 64];
+    impl_random_arrays![4 8 12 16 24 32 48 64];
 }
 
 /// A type that can be returned by `ring::rand::generate()`.
```

**Comment by briansmith** [comment]
Thanks. Regarding getting the code into master: There are multiple *ring* repos. Accidentally the default upstream for `master` on my local tree had been changed to another upstream and so I pushed the commits for the 0.16.8 release there, instead of to this repo, by accident.

I am surprised this is a breaking change because I'd tested this with some projects before the release specifically to avoid this problem. However, I did not test `biscuit`. Indeed, if I try to build `biscuit` with 0.16.8 now with the updated *ring* it fails to build.

I've released 0.16.9 with the `LessSafeKey` changes reverted. I yanked 0.16.8. The code is here in this repo in https://github.com/briansmith/ring/commit/56cddd73d20fc440007c893cbd49d9e7633ae697.

I will release a new version with the new AEAD random nonce functionality added in a different, non-breaking way.
